### PR TITLE
CDAP-5186 make new etl config backwards compatible

### PR DESCRIPTION
**Pull request by albertshau**
Adds the old v1 fields (source, transforms, sinks) to the v2
configs, and add a method to convert those fields into the
stages field.

With that, we can switch the apps to the v2 configs and remove
the old PipelineRegisterer logic that has been replaced with the
PipelineSpecGenerator. This also makes it so that the apps can
now understand the new style configs, which sets up implementation
of the smart workflow and aggregator plugin types.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT3730

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineTypeValidator.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+import java.lang.reflect.GenericArrayType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.List;
+
+/**
+ * Validates types passed between pipeline stages.
+ */
+public class PipelineTypeValidator {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineTypeValidator.java]: this class is the remains of the PipelineRegisterer

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/ETLBatchApplication.java [hunk 2]
```diff
@@ -38,26 +48,33 @@
 public class ETLBatchApplication extends AbstractApplication<ETLBatchConfig> {
   public static final String SCHEDULE_NAME = "etlWorkflow";
   public static final String DEFAULT_DESCRIPTION = "Extract-Transform-Load (ETL) Batch Application";
+  private static final Set<String> SUPPORTED_PLUGIN_TYPES = ImmutableSet.of(
+    BatchSource.PLUGIN_TYPE, BatchSink.PLUGIN_TYPE, Transform.PLUGIN_TYPE,
+    Constants.CONNECTOR_TYPE, BatchAggregator.PLUGIN_TYPE);
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/ETLBatchApplication.java]: We should not have connector and batch aggregator for ETL Batch. It should be for data pipeline.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineTypeValidator.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineTypeValidator.java]: nit. 2016

File: cdap-app-templates/cdap-etl/cdap-etl-realtime/src/main/java/co/cask/cdap/etl/realtime/ETLWorker.java [hunk 4]
```diff
@@ -122,13 +132,24 @@ public void configure() {
...
+                                RealtimeSink.PLUGIN_TYPE,
+                                Table.class,
+                                DatasetProperties.builder()
+                                  .add(Table.PROPERTY_SCHEMA, ERROR_SCHEMA.toString())
+                                  .build());
+    PipelineSpec spec = specGenerator.generateSpec(config);
+    PipelinePlanner planner = new PipelinePlanner(SUPPORTED_PLUGIN_TYPES, ImmutableSet.<String>of());
+    PipelinePlan plan = planner.plan(spec);
+    if (plan.getPhases().size() > 1) {
+      throw new IllegalArgumentException("Only one source is supported at this time.");
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-realtime/src/main/java/co/cask/cdap/etl/realtime/ETLWorker.java]: Do you mean `Only one phase is supported at this time.`?
Or is there the correlation defined that the number of phases corresponds to the number of sinks?

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-realtime/src/main/java/co/cask/cdap/etl/realtime/ETLWorker.java]: yeah there is a one-to-one correlation.  But I will change it to check directly so its more clear and less brittle.

**Comment by sreevatsanraman** [comment]
LGTM

**Pull request merged**